File: python/stack/binary-tree-postorder-traversal[2].py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def postorderTraversal(self, root):
        st = []
        data = []
        pre = None

        if root:
            st.append(root)
            while st:
                cur = st[len(st)-1]
                if ((cur.left == None and cur.right == None) or \
                    (pre  and (pre == cur.left or pre == cur.right))):
                    data.append(cur.val)
                    st.pop()
                    pre = cur
                else:
                    if cur.right:
                        st.append(cur.right)
                    if cur.left:
                        st.append(cur.left)

        return data

File: go/stack/binary-tree-postorder-traversal[2].go
package stack

func postorderTraversal2(root *TreeNode) []int {
	var result []int

	if root == nil {
		return result
	}

	st := New()
	st.Push(root)
	var pre *TreeNode
	for !st.IsEmpty() {
		cur := st.Top().(*TreeNode)
		if (cur.Left == nil && cur.Right == nil) ||
			(pre != nil && (cur.Left == pre || cur.Right == pre)){
			result = append(result, cur.Val)
			st.Pop()
			pre = cur
		}else {
			if cur.Right != nil {
				st.Push(cur.Right)
			}

			if cur.Left != nil {
				st.Push(cur.Left)
			}
		}
	}

	return result
}

